pkg/types/spec/errors.go

```diff
@@ -222,45 +273,207 @@ func ErrorPythonPathNotFound(pythonPath string) error {
 func ErrorS3FileNotFound(path string) error {
 	return errors.WithStack(&errors.Error{
 		Kind:    ErrS3FileNotFound,
-		Message: fmt.Sprintf("%s: not found or insufficient permissions", path),
+		Message: fmt.Sprintf("%s: file not found or insufficient permissions", path),
+	})
+}
+
+func ErrorS3DirNotFound(path string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrS3DirNotFound,
+		Message: fmt.Sprintf("%s: dir not found or insufficient permissions", path),
+	})
+}
+
+func ErrorS3DirIsEmpty(path string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrS3DirIsEmpty,
+		Message: fmt.Sprintf("%s: S3 directory is empty", path),
+	})
+}
+
+func ErrorModelPathNotDirectory(modelPath string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrModelPathNotDirectory,
+		Message: fmt.Sprintf("%s: model path must be a directory", modelPath),
+	})
+}
+
+var _pythonVersionedExpectedStructMessage = `
+  %s
+  ├── 1523423423/ (Version prefix)
+  |   └── * // Model-specific files (i.e. model.h5, model.pkl, labels.json, etc)
+  └── 2434389194/ (Version prefix)
+	  └── * // Model-specific files (i.e. model.h5, model.pkl, labels.json, etc)
+`
+
+var _pythonExpectedStructMessage = `
+  %s
+  └── * // Model-specific files (i.e. model.h5, model.pkl, labels.json, etc)
+`
+
+func ErrorInvalidPythonModelPath(modelPath string, s3Path bool, includeVersionMessage bool, includeNonVersionMessage bool, modelSubPaths []string) error {
```

Comment:
We can probably add comments inside each function to describe the process. I wouldn't break it up into multiple errors because that will lead to an excess of error functions that will be confusing to track from the calling functions. Even more so, there wouldn't be any semantic difference between the different error functions, which I think, will lead to even more confusion.

The reason of only having a single error function for each predictor type is that regardless of the provided model, if something is bad with the model's structure, then that will be observed in the side-by-side comparison between the template model and the provided one. I think the side-by-side comparison is an efficient way of describing the bug, because it reduces the number of error functions we need and because the user can immediately see what's wrong with the provided model.

Here's an example for the ONNX predictor:
```text
s3://cortex-0/models/iris-classifier: invalid ONNX model path. For models provided for the ONNX predictor type, the path must be a directory with the following structure:
models/iris-classifier/
  ├── 1523423423/ (Version prefix)
  |   └── <model-name>.onnx // ONNX-exported file
  └── 2434389194/ (Version prefix)
      └── <model-name>.onnx // ONNX-exported file

or like

models/iris-classifier/
└── <model-name>.onnx // ONNX-exported file

but its current structure is

models/iris-classifier/678/
└── model.bin
```

Or like this:
```text
s3://cortex-0/models/iris-classifier: invalid ONNX model path. For models provided for the ONNX predictor type, the path must be a directory with the following structure:
models/iris-classifier/
  ├── 1523423423/ (Version prefix)
  |   └── <model-name>.onnx // ONNX-exported file
  └── 2434389194/ (Version prefix)
      └── <model-name>.onnx // ONNX-exported file

or like

models/iris-classifier/
└── <model-name>.onnx // ONNX-exported file

but its current directory is empty
```

We should probably discuss this live.

Reply:
Done the Python one. Idem for the ONNX/TensorFlow error functions.